### ISSUE TITLE
A Compose wrapper for lr scheduler

### ISSUE THREAD

**Issue by harryhan618** [pytorch/pytorch#41307]
## 🚀 Feature
A `Compose` class for lr scheduler like `torchvision.transforms.Compose`. This `Compose` is initialized with a list. In the list, each item is a list: [single_lr_scheduler, start_iter_idx, end_iter_idx]. 

## Motivation

In pytorch 1.4, chaining lr scheduler is supported. Usually a warmup strategy is used. With a `Compose`, one single `step` can be called in a neat way. 

cc @vincentqb

**Comment by malfet** [comment]
@harryhan618 thank you for your suggestion. If you have patch in mind that addresses this problem, please do not hesitate to publish.

**Comment by harryhan618** [comment]
@malfet Hi, may I ask a question? Why do we set `last_epoch` in `init` as -1 and call `step` in `init`? Why don't we set `last_epoch` to be 0 directly?
Because this involves the corner case of `last_epoch=0`. Some schedulers update when `last_epoch=0`, e.g. `LambdaLR`. Others don't, e.g. `ExponentialLR`. This corner case happens when `step` is called in `init`. So the behavior is kinda weird.

**Comment by vincentqb** [comment]
A compose wrapper should not be needed. Instead, one should simply be able to chain them. See #26423.

**Comment by harryhan618** [comment]
@vincentqb But in #26423, it seems we have to call `step` for every lr_scheduler. But when therea are multiple lr_schedule strategies, we have to judge iteration_idx / epoch_idx to see whether we should should a `step`. 

Also, there is an issue #37768 and corresponding RP #37770, discussing `step` by epoch or by iteration .

So I'm thinking a `Compose` wrapper so that we can call `step` once right after the `optim.step()`, like this:
``` python
lr_comp = LRCompose_Init_API   # a list of lr_schedulers, either by epoch or by iteration, also with their start time and end time
...
for x, y in dataloader:
    ...
    optim.step()
    lr_comp.step()
```
We can convert those `step`-by-epoch operations to iteration index to decide whether to call them. Thus a fine grained lr scheduler can be called in a neat way.

**Comment by vincentqb** [comment]
> @vincentqb But in #26423 , it seems we have to call `step` for every lr_scheduler. But when there are multiple lr_schedule strategies, we have to judge iteration_idx / epoch_idx to see whether we should call a `step`.

Yes, a user would need to call step for each optimizer they want to step. This offers flexibility to let the user combine them however they want. Is there another disadvantage you notice?

By the way, a similar proposal was discussed [here](https://github.com/pytorch/pytorch/pull/21800#issuecomment-512291714).

> Also, there is an issue #37768 , discussing `step` by epoch or by iteration .

The distinction of epoch and iteration depends somewhat on the point of view of the user and the application in mind. The scheduler doesn't see that. It only updates a value every time `step` is called. In particular, this means users can choose the parameters of the scheduler to match whether they are looking at "epoch" or "iterations".

> So I'm thinking a `Compose` wrapper so that we can call `step` like this:
> 
> ```python
> """
> we may have a bunch of lr_schedulers, with warmup strategy, one lr_scheduler in the middle of training, another in the end, another by epoch. Or, some lr_schedulers for one part of the model during some time, others on other part of the model, etc.
> """
> lr_comp = LRCompose_Init_API   # a list of lr_schedulers, either by epoch or by iteration, also with their start time and end time
> ...
> for x, y in dataloader:
>     ...
>     optim.step()
>     lr_comp.step()
> lr_comp.step()
> ```
> 
> No need to judge if lr_scheduler is by iteration or by epoch explicitly for users.

It looks to me like the main gain here is saving on the number of times a user type `step`. What do you think?  When writing the composed scheduler, the user would still need to decide whether each is stepped at an iteration or an epoch.

**Comment by harryhan618** [comment]
@vincentqb Hi. I think we both agree that we should leave the flexibility to users.
My suggestion is that, with a Compose, a user can manage different LR_Scheduler at different without iter_idx / epoch_idx judgement. So the code for training procedure would be more simple and self-explained.
But it seems currently, it is not a common practice for users to use complicated lr_scheduler strategy. Also, the usage of by iteration and by epoch makes it complicated and I haven't come up a neat and concise API design. So I'm going to close this issue and its corresponding PR for now. 
Please feel free to talk to me if you have more ideas. Thank you!
